models/contracts/optimism/contracts_optimism_contract_mapping.sql

```diff
@@ -12,8 +12,8 @@
   )
 }}
 
--- set max number of levels to trace root contract
-{% set max_levels = 5 %}
+-- set max number of levels to trace root contract, eventually figure out how to make this properly recursive
+{% set max_levels = 10 %}
```

Comment:
Do you know any indicator or feature a root contract may have? It seems this spell is a bit logically complicated, I will try to see if I can help to simplify it a bit.

Reply:
If by this you mean when to "break", it's when u.contract_address is null

Reply:
It is hard to precisely set max_level and other termination conditions, since spark sql does not support recursive CTE at the moment.

There is also a spell `labels_contract_deployers_optimism` which might be useful here.

Reply:
i'll bump back down to ~7 for now. Maybe a later iteration we can explore exit times